clash-cores/src/Clash/Cores/Xilinx/DcFifo/BlackBoxes/Internal.hs

```diff
@@ -0,0 +1,217 @@
+{-# LANGUAGE RecordWildCards #-}
+{-# LANGUAGE ViewPatterns #-}
+
+module Clash.Cores.Xilinx.DcFifo.BlackBoxes.Internal where
+
+import Prelude
+
+import Clash.Core.Literal (Literal(NaturalLiteral))
+import Clash.Core.TermLiteral (termToDataError, TermLiteralSNat(..))
+import Clash.Core.Term (Term(Literal))
+import qualified Clash.Primitives.DSL as DSL
+import Clash.Netlist.BlackBox.Types (BlackBoxFunction, emptyBlackBoxMeta)
+import Clash.Netlist.Types (TemplateFunction(..), BlackBox(BBFunction))
+import Clash.Netlist.Util (orNothing)
+
+import qualified Clash.Netlist.Id as Id
+import qualified Clash.Netlist.Types as N
+import qualified Clash.Netlist.BlackBox.Types as N
+
+import Control.Monad (unless)
+import Data.Either (lefts)
+import Data.Maybe (catMaybes)
+import GHC.Stack (HasCallStack)
+
+import Clash.Cores.Xilinx.DcFifo.Implementation (DcConfig(..))
+import Clash.Cores.Xilinx.DcFifo.Instances ()
+import Clash.Cores.Xilinx.Internal (toTclBool, renderTcl, defIpConfig, IpConfig (properties))
+
+-- | Blackbox function for 'Clash.Cores.Xilinx.Fifo.dcFifo'. It parses the "DcConfig"
+-- supplied to 'dcFifo' from its Term representation, and passes them to two
+-- template functions:
+--
+--   * 'dcFifoTclTF': renders TCL file calling Xilinx's `create_ip`
+--   * 'dcFifoTF': instantiates IP generated in 'dcFifoTclTF'
+--
+-- Additionally, it generates a unique module name for the Xilinx IP.
+--
+dcFifoBBF :: HasCallStack => BlackBoxFunction
+dcFifoBBF _isD _primName args _resTys = do
+  let
+    [  _knownDomainWrite, _knownDomainRead, knownNatN
+     , _bitPack, _, _nfDataX, _knownNatDepth
+     , _constraint1, _constraint2
+     , either error id . termToDataError -> dcConfig
+     , _wClk, _rClk, _rst, _wData
+     , _rEnable
+
+    -- TODO: Make this blackbox return multiple results, instead of a tuple. See:
+    --       https://github.com/clash-lang/clash-compiler/pull/1560
+    --  , _, _, _, _, _, _, _
+     ] = lefts args
+
+  let width =
+        case knownNatN of
+          Literal (NaturalLiteral n) -> fromInteger n
+          _ -> error "Unexpected type of knownNatN"
+
+  pure (Right (bbMeta width dcConfig, bb dcConfig))
+ where
+  bbMeta width dcConfig = emptyBlackBoxMeta
+    { N.bbKind = N.TDecl
+    , N.bbIncludes =
+      [ ( ("dcfifo", "tcl")
+        , BBFunction (show 'dcFifoTclTF) 0 (dcFifoTclTF width dcConfig))
+      ]
+    -- TODO: Make this blackbox return multiple results, instead of a tuple. See:
+    --       https://github.com/clash-lang/clash-compiler/pull/1560
+    -- , N.bbResultNames =
+    --   , N.BBTemplate [N.Text "wr_full"]
+    --   , N.BBTemplate [N.Text "wr_data_count"]
+
+    --   , N.BBTemplate [N.Text "rd_empty"]
+    --   , N.BBTemplate [N.Text "rd_data_count"]
+    --   , N.BBTemplate [N.Text "rd_dout"]
+    --   ]
+    }
+
+  bb :: DcConfig TermLiteralSNat -> BlackBox
+  bb dcConfig = BBFunction (show 'dcFifoTF) 0 (dcFifoTF dcConfig)
+
+-- | Instantiate IP generate with 'dcFifoTclTF'.
```

Comment:
```suggestion
-- | Instantiate IP generated with 'dcFifoTclTF'
```

Reply:
I think all the spelling mistakes are my sloppy work, sorry @vmchale